### ISSUE TITLE
[27] Cross-Domain Few-Shot Classification via Learned Feature-Wise Transformation

### ISSUE THREAD

**Issue by long8v** [long8v/PTIR#32]
<img width="847" alt="image" src="https://user-images.githubusercontent.com/46675408/169812425-e2791252-5faa-4709-a7fe-cc71509fcb5f.png">

[paper](https://arxiv.org/pdf/2001.08735.pdf), [code](https://github.com/hytseng0509/CrossDomainFewShot)

## TL;DR 
**problem :** few-shot classification은 같은 domain(=ImageNet 내에서 unseen label을 예측)에서는 잘 작동되지만, 다른 도메인으로 few-shot을 할 경우 잘 작동되지 않음(ImageNet으로 훈련된게 CUB 데이터로 few-shot test를 하면 잘 안나옴)
**solution :** feature encoder에 feature-wise transformation layer(affine 변환)를 추가하였고 이때의 하이퍼파라미터는 learning-to-learn 방법론으로 학습됨. 
**result :** MatchingNet, RelationNet, Graph Neural Network에 위의 feature-wise transformation을 적용했을 때 generalization 성능이 좋았음. 

## details 
- domain adaption / generalization의 차이는 generalization의 경우 학습 단계에서 unseen domain을 사용하지 않고 generalize 해야 함.
- 우리는 domain generalization 문제를 few-shot 셋팅에서 novel한 클래스를 분류하는 문제로 바꿈. 

### 3.1. Preliminaries
- few-shot terms
  - N_w : # of categories
  - N_s : # of labeled examples for each categories
- 아래 그림은 3 way 3 shot few shot의 예시
<img width="813" alt="image" src="https://user-images.githubusercontent.com/46675408/169812728-1c1dc722-3752-48da-9607-1f66aa0ca7b1.png">

- metric-based의 알고리즘은 feature encoder E와 metric function M으로 구성되어 있음. 
- 각 iteration에서 N_w개의 카테고리를 뽑고 task T를 만든다. input image를 X, 이에 해당하는 label을 Y라고 하고. task T는 support set인 S={(X_s, T_s)}와 query set인 {(X_q, Y_q)}로 구성된다. 
- feature encoder E는 support와 query 이미지의 feature를 뽑고 metric function M에 넣어 support image의 label을 참고하여 query image의 카테고리를 예측한다. 
<img width="326" alt="image" src="https://user-images.githubusercontent.com/46675408/169816949-89019fb0-b3af-4ba5-bc43-fb24b3ab5152.png">

- 학습 목표 함수는 query set에 대한 이미지의 classification loss이다. 
<img width="251" alt="image" src="https://user-images.githubusercontent.com/46675408/169816998-b546aedb-6ff3-46ae-926a-34608093d151.png">

- 다양한 metric-based 알고리즘의 주요한 차이점은 이미지 피쳐를 뽑는 E의 아키텍쳐에 따라 달라진다. 가령 MatchingNet은 LSTM, RelationNet은 CNN, GNN은 GCN
- 학습할 때 seen domain들로 training하고 평가는 unseen domain에 대하여 하였다.

### 3.2. feature-wise transformation layer
<img width="251" alt="image" src="https://user-images.githubusercontent.com/46675408/169821093-45bd57ed-8478-48f3-805d-369e391578ab.png">

- 우리의 목표는 unseen 데이터에 대해 generalization을 더 잘하기 위함인데, metric function M이 seen domain에 overfitting되기 쉬우므로 이를 막아줘야한다.
- 직관적으로, feature encoder E에 affine tranformations을 적용하면 더 다양한 분포를 표현할 수 있을 것 같다.
- hyper parameter는 아핀 변환 파라미터를 sampling 하기 위한 standard dev를 뜻한다.
<img width="536" alt="image" src="https://user-images.githubusercontent.com/46675408/169819651-da38dd31-7527-4743-8f21-2ffba623de0b.png">

- batch norm 이후에 아래의 feature-wise transformation layer를 적용한다. 
<img width="324" alt="image" src="https://user-images.githubusercontent.com/46675408/169819934-5fd871b7-12fe-47ba-a991-bf2a5224bb5b.png">

### 3.3 Leraning the feature-wise transformation layers(=FT layer)
- 위의 하이퍼파라미터를 경험적으로 선택할 수도 있겠지만 학습할 수 있도록 하고 싶다. 우리는 이를 위해 learning-to-learn 알고리즘을 디자인했다. 주요 아이디어는 FT 레이어를 적용하여 seen domain에 대해서 학습한 것이 unseen domain에 대해서도 나은 성능을 내게 하는 것이다.
<img width="734" alt="image" src="https://user-images.githubusercontent.com/46675408/169821123-b9ebe757-03ab-4ac2-bd11-371093025a17.png">

...작성 중 

- but  나의 문제는 sequence 문제. 흠.. -> cross domain NER로 검색을 하는게 나을듯.
